src/webview/js/js.js

```diff
@@ -442,7 +499,14 @@ const sendScrollMessage = () => {
     endMessageId: rangeHull.last,
   });
   setMessagesReadAttributes(rangeHull);
-  prevMessageRange = messageRange;
+  // If there are no visible + read messages (for instance, the entire screen
+  // is taken up by a single large message), then we don't want to update
+  // prevMessageRange, since doing so could cause unread messages that are
+  // above the currently visible message to not get marked read next time
+  // sendScrollMessage is called.
```

Comment:
This is accurate but can perhaps be a bit clearer by describing directly how it does behave instead of how it avoids behaving:
```suggestion
  // prevMessageRange.  This way, if the user scrolled past some messages
  // to get here, then even though `messageRange` was empty this time and so
  // we didn't mark any messages as read just now, we'll include those in `rangeHull`
  // the next time the user scrolls and so we'll mark them as read then.
```

(Except wrapped properly -- that's more annoying to do in the GitHub comment form than in a full editor.)

Writing it out that way, I also get the feeling that this is a bit of a kludgy way of handling this case: it'd be better to go ahead and mark now the messages we *do* already know we want to mark as read. So in particular if the user scrolled down to get here, that means the messages from wherever they were up through the message that's just above the current visible one.

But getting that right may open up a new set of subtleties to debug, so it's probably best left for a followup. I think this behavior is good enough that it doesn't need to block merging this PR and getting the benefits of the main change it makes.

Reply:
Yeah, I agree that this is kludgy, but it seemed like the most likely to be correct thing. In general, it seems like it would be nice to rewrite a lot of this code to use a more abstract model of messages (ie, not directly operate on the DOM), so that it would be more easy to test.

Reply:
Let's continue discussion in [the chat thread](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/marking.20read.20on.20scroll/near/1155043).